streaming/sender.c

```diff
@@ -715,3 +733,122 @@ void *rrdpush_sender_thread(void *ptr) {
+    RRDDIM *rd;
+    struct rrddim_query_handle handle;
+
+    // When the sync is unknown against the far end (last_sent=0), sending the latest sample will trigger a
+    // replication from the far end if necessary.
+    time_t first_t = rrdset_first_entry_t(st);
+    time_t st_newest = st->last_updated.tv_sec;
+    time_t window_start;
+
+    if (start_time == 0)
+        window_start = MAX((time_t)st->state->last_sent.tv_sec + st->update_every, first_t);
+    else
+        window_start = MAX(start_time, first_t);
+
+    // Decide how much data to replicate from the beginning of the window
+    time_t unsent_points = (st_newest - window_start) / st->update_every + 1;
+    if (unsent_points > default_rrdpush_gap_block_size)
+        unsent_points = default_rrdpush_gap_block_size;
+    time_t window_end = window_start + unsent_points * st->update_every;    // window_end <= st_newest + update_every
+
+    // If we are responding to an explicit request then we may send empty windows, these communicate to the far end
+    // that there is no data in the interval. There are times in the REPBEGIN:
+    //   * The beginning of the window in time (e.g. where we are reporting from including gaps)
+    //   * The beginning of the data in time   (e.g. dropping leading gaps from the window)
+    //   * The end of the window.
+    // The start times are inclusive and the end time is exclusive, so if the requested window is 10..20 and the
+    // chart contains data at times 15,16,17,18 then the response will be REPBEGIN 10 15 19
+    if (start_time == 0)
+        buffer_sprintf(s->build, "REPBEGIN %s %ld %ld %ld\n", st->id, window_start, window_start, window_end);
+    else
+        buffer_sprintf(s->build, "REPBEGIN %s %ld %ld %ld\n", st->id, start_time, window_start, window_end);
+
+    size_t num_points = 0;
+    rrddim_foreach_read(rd, st) {
+        // Send the intersection of this dimension and the time-window on the chart
+        if (!rd->exposed)
+            continue;
+        time_t rd_start = rrddim_first_entry_t(rd);
+        time_t rd_end   = rrddim_last_entry_t(rd) + st->update_every;
+        if (rd_start < window_end && rd_end >= window_start) {
+
+            time_t rd_oldest = MAX(rd_start, window_start);
+            rd_end = MIN(rd_end,   window_end);
+
+            rd->state->query_ops.init(rd, &handle, rd_oldest, rd_end);
+            debug(D_REPLICATION, "Fill replication with %s.%s @%ld-%ld rd@%ld-%ld", st->id, rd->id, window_start,
+                                 window_end, rd_oldest, rd_end);
+
+            for (time_t metric_t = rd_oldest; metric_t < rd_end; ) {
+
+                if (rd->state->query_ops.is_finished(&handle)) {
+                    debug(D_REPLICATION, "%s.%s query handle finished early @%ld", st->id, rd->id, metric_t);
+                    break;
+                }
+
+                storage_number n = rd->state->query_ops.next_metric(&handle, &metric_t);
+                if (n==SN_EMPTY_SLOT)
```

Comment:
please add spaces between comparison and variables.

Reply:
Fixed.